File: mobile/lib/features/workouts/domain/weather_condition.dart
import 'package:flutter/material.dart';

import '../../../l10n/app_localizations.dart';

/// The fixed set of weather condition codes the app's own manual-entry
/// picker offers (docs/cardio/60 C8.6, Q-C8.1) — `CardioMetrics.weatherCondition`
/// itself is a free, unconstrained string on the wire (same precedent as
/// `gameFormat`), so an older/newer client sending a code outside this list
/// is never a parse error, just an icon this client doesn't recognize
/// ([weatherConditionIcon] falls back to [Icons.cloud_off] for it).
const List<String> kWeatherConditions = [
  'CLEAR',
  'PARTLY_CLOUDY',
  'CLOUDY',
  'RAIN',
  'SNOW',
  'WINDY',
];

/// Material icon for a weather condition code — matches
/// `design/Lifey Cardio Sport-specifikumok.dc.html` M42 (`partly_cloudy_day`
/// there maps to [Icons.wb_cloudy] here, the closest bundled equivalent).
/// An unrecognized/null code — including "no weather data at all" — gets
/// the same [Icons.cloud_off] the M42 empty state itself uses.
IconData weatherConditionIcon(String? code) {
  return switch (code) {
    'CLEAR' => Icons.wb_sunny,
    'PARTLY_CLOUDY' => Icons.wb_cloudy,
    'CLOUDY' => Icons.cloud,
    'RAIN' => Icons.water_drop,
    'SNOW' => Icons.ac_unit,
    'WINDY' => Icons.air,
    _ => Icons.cloud_off,
  };
}

/// Localized label for a weather condition code.
String weatherConditionLabel(AppLocalizations l10n, String code) {
  return switch (code) {
    'CLEAR' => l10n.weatherConditionClear,
    'PARTLY_CLOUDY' => l10n.weatherConditionPartlyCloudy,
    'CLOUDY' => l10n.weatherConditionCloudy,
    'RAIN' => l10n.weatherConditionRain,
    'SNOW' => l10n.weatherConditionSnow,
    'WINDY' => l10n.weatherConditionWindy,
    _ => code,
  };
}

File: web/src/features/workouts/weatherCondition.ts
/**
 * The web port of the mobile `weather_condition.dart` — `cardio.weatherCondition`
 * is a free, unconstrained string on the wire (same precedent as `gameFormat`),
 * so an unrecognized code is never a parse error, just an icon/label this
 * client falls back on.
 */

const ICON_BY_CODE: Record<string, string> = {
  CLEAR: "wb_sunny",
  PARTLY_CLOUDY: "wb_cloudy",
  CLOUDY: "cloud",
  RAIN: "water_drop",
  SNOW: "ac_unit",
  WINDY: "air",
};

const LABEL_KEY_BY_CODE: Record<string, string> = {
  CLEAR: "cardioWeatherConditionClear",
  PARTLY_CLOUDY: "cardioWeatherConditionPartlyCloudy",
  CLOUDY: "cardioWeatherConditionCloudy",
  RAIN: "cardioWeatherConditionRain",
  SNOW: "cardioWeatherConditionSnow",
  WINDY: "cardioWeatherConditionWindy",
};

/** Material Symbols icon name — `cloud_off` for null/unrecognized, matching mobile's fallback. */
export function weatherConditionIcon(code: string | null): string {
  return (code && ICON_BY_CODE[code]) || "cloud_off";
}

/** The i18n key for a condition code, or `null` for an unrecognized one — the caller shows the raw code then. */
export function weatherConditionLabelKey(code: string): string | null {
  return LABEL_KEY_BY_CODE[code] ?? null;
}
